services/notification/notifiers/mixins/message/__init__.py

```diff
@@ -202,98 +201,92 @@ def _possibly_write_install_app(
-                write(line)
+            wrote_something |= line is not None
+            write(line)
+
         if wrote_something:
             write("")
 
-    def get_middle_layout_section_names(self, settings):
-        sections = map(
-            lambda layout: layout.strip(), (settings["layout"] or "").split(",")
-        )
-        return [
-            section
-            for section in sections
-            if section
-            not in [
-                "header",
-                "newheader",
-                "newfooter",
-                "newfiles",
-                "condensed_header",
-                "condensed_footer",
-                "condensed_files",
-            ]
-        ]
-
-    def get_upper_section_names(self, settings):
-        sections = list(
-            map(lambda layout: layout.strip(), (settings["layout"] or "").split(","))
-        )
-        headers = ["newheader", "header", "condensed_header"]
-        if all(x not in sections for x in headers):
+    def get_middle_layout_section_names(self, settings: dict) -> list[str]:
+        filtered_sections = {
+            "header",
+            "newheader",
+            "newfooter",
+            "newfiles",
+            "condensed_header",
+            "condensed_footer",
+            "condensed_files",
+        }
+
+        sections = get_sections(settings)
+
+        return [section for section in sections if section not in filtered_sections]
+
+    def get_upper_section_names(self, settings: dict) -> list[str]:
+        headers = {"newheader", "header", "condensed_header"}
+        allowed_sections = {
+            "header",
+            "newheader",
+            "condensed_header",
+            "newfiles",
+            "condensed_files",
+        }
+
+        sections = get_sections(settings)
+
+        if set(sections).isdisjoint(headers):
             sections.insert(0, "condensed_header")
 
         if "files" in sections or "tree" in sections:
             sections.append("newfiles")
 
-        return [
-            section
-            for section in sections
-            if section
-            in [
-                "header",
-                "newheader",
-                "condensed_header",
-                "newfiles",
-                "condensed_files",
-            ]
-        ]
-
-    def get_lower_section_name(self, settings):
-        if (
-            "newfooter" in settings["layout"]
-            or "condensed_footer" in settings["layout"]
-        ):
+        return [section for section in sections if section in allowed_sections]
+
+    def get_lower_section_name(self, settings: dict) -> str | None:
+        sections = get_sections(settings)
+
+        if "newfooter" in sections or "condensed_footer" in sections:
             return "newfooter"
+        return None
+
+
+def get_sections(settings: dict) -> list[str]:
+    layout = (settings.get("layout") or "").strip()
+    if not layout:
+        return []
```

Comment:
I think if you return nothing the PR comment would be empty.
It's better to return the default value in case there is none.

The default value is [here](https://github.com/codecov/shared/blob/fccd4796bfb4df48eb7c6f17191c145c3c587109/shared/config/__init__.py#L36) (and it should not be lost when the YAML is processed, but merged with the user's option - I think that's the real bug)